README.md
Baidu Speech to Text
=====================

Latest Version: 2.0.2 [CHANGELOG](CHANGELOG.md)

Hi, this project is aim to speech to text by baidu ai service, now, it's good for chinese
you can also use it for English.

you can send me email [aokihu@gmail.com](mailto:aokihu@gmail.com)

# BIG CHANGE 💣

Because I use `snowboy` to support hot word, so it is only running on Nodejs **6.0** now.

Before you install the module, you must install `sox` on Mac OS X or Windows, or `arecord` on Linux
and `alsa` on Linux

# Prepare

The project depends [mic](https://www.npmjs.com/package/mic), so you should install **sox**(Mac/Windows) OR **arecord**(Linux) before install this module.

you can install **arecord** under linux with command, **arecord** is apart of **alsa-utils**

```bash
$ sudo apt-get update
$ sudo apt-get upgrade
$ sudo apt-get install alsa-base alsa-utils
```

you can install **sox** on Mac with [**brew**](https://brew.sh/)

```bash
brew install sox
```

If you want have continual ablitilty, you should **USE Node.js 6.0 OR LTS Version product**, and you can find more detial information with [Snowboy](https://www.npmjs.com/package/snowboy)


# Install 

you can use `npm`

```bash
npm install baidu-stt
```

or you can also use yarn

```bash
yarn add baidu-stt
```

After install the module, you should make a `resources` folder and copy resources file to the folder
```bash
mkdir ./resources
cp ./node_modules/baidu-stt/resources/Alexa.pmdl ./resources
```


# How to use

```javascript
const Baidu_STT = require('baidu-stt');

const bdstt = new Baidu_STT({
  apiKey: '......',
  secertKey: '......',
  language:'zh',
  recordVoice:false,
  debug:false,
  continual:false,
  modelFile:'./resources/Alexa.pmdl',
  sensitivity:'0.5',
  gain:2.0,
  silenceDelay:2000,
  hotwords:"Alexa",
  voicePath:'./',
  voiceRate:'16000',
  voiceType:'wav'
})

bdstt.init();
bdstt.on('ready', () => {
  console.log('Start listen...');
  bdstt.listen();
});

bdstt.on('start', () => console.log('Please speak something, I\'m listening...'));
bdstt.on('listening', () => console.log('I\'m listen'));
bdstt.on('stop', () => console.log('Copy that'));
bdstt.on('upload', () => console.log('Uploading voice data...'));

bdstt.on('success', data => {
  console.log(data);
});

bdstt.on('fail', err => console.log(err));
```

**HOT WORD**

The default hot word is `Alexa`, for it is the most popular keywords on snowboy, you can train your personal hot word with [snowboy website](https://snowboy.kitt.ai/)

Becareful, personal model file extesion is `.pdml`, it is avalibe file type, you can directly use it.

# API

**Baidu_STT(options)**

Return an Baidu_STT object instance

- **options** JSON containing command line options. Following are valid options:
  - apiKey ***{string}*** App api key
  - secretKey ***{string}*** App secret key
  - language ***{stirng}*** Choose your language, default is `'zh'`, you can choose `'en'` for `English`
  - continual ***{string}*** Continue to recoding after uploading voice data, if `true` it will be contunal
  - modelFile ***{string}*** The hot word detected model file, default is ***./resources/Alexa.pmdl***, you should copy the model file to ***./resources*** folder
  - hotwords ***{string}*** The keyword for wakeing up, default is ***Alexa***
  - sensitivity ***{string}*** 
  - debug ***{string}*** Display capture data
  - recordVoice ***{boolean}*** Whether save the voice on your disk, default is `false`
  - voiceRate ***{string}*** The rate of voice record, default is `'16000'`, you can set `'8000'` or any vaild vaule in `sox` OR `arecord`
  - voicePath ***{string}*** The path of saved voice, default is `./`
  - voiceType ***{string}*** The type of recod voice file, default is `wav`

**instance.init()**

Request token then emit `ready` event;

**instance.listen()**

Set instance listen for your voice;

**instance.pause()**

When you set contiunal is `true`, you can use this fucntion pause record sound;

**instance.resume()**

When you set continual is `true` and paused, you can use it to resume;

**instance.stop()**

Stop recording.


# Event

**start**

Start record voice after you call `listen()`

**wakeup**

Be waked up by hot word

**sleep**

After upload voice data, it is going to sleep status

**listening**

Recoding voice

**stop**

Stop recoding voice

**upload**

Upload voice date to service

**success**

Success for recogniting your voice content, and return the result
it is `array` type, because there may be many result. Normally the first result is better than other. I see

**fail**

When baidu can not recognting, it will return error message, it's `string` type.


CHANGELOG.md
CHANGE LOG
===========

* Version 2.0.2

Add 3 function
> stop()
> pause()
> resume()

* Version 2.0.1

Just update README

* Version 2.0.0

**BREAK CHANGE**

Now, you just only use node.js 6.0, because IMPORT `snowboy` module, it support hot word to wake up, so you can continual speak, without restart.

You can say `Alexa` to wake it up, OR you also train your custom voice command on [Snowboy Site](https://snowboy.kitt.ai/)

* Version 1.0.4

Add new construct param `continual`, if it is `true` it will continual recording, if is `false` it will stop after upload voice data

* Version 1.0.3

Add new construct param `voiceRate`, you can change the rate by yourself, default is `'8000'`, ***ATTATION*** the param's type is `***STRING***`

* Version 1.0.1

Add new event `fail`, when baidu service can not recongniting , it will return error message , type is `string`

* Version 1.0.0

***Publish***


test.js
/**
 * Test file
 * @version 2.0.1
 */

const BaiduSTT = require('./index.js');

const bdstt = new BaiduSTT({
  apiKey: 'DAZTeCYAkuAQwtPvWDcBb9Ih',
  secretKey: 'adb28deb5103bde96bd9633dd437ff2d',
  voiceRate: '16000',
  // debug:true,
  continual: true,
  sensitivity: "0.4",
  modelFile:'./resources/Alexa.pmdl',
  hotwords:'Alexa'
 });

bdstt.init();
bdstt.on('ready', () => {
  console.log('Start listen...');
  bdstt.listen();
});

bdstt.on('start', () => console.log('Please speak something, I\'m listening...'));
// bdstt.on('listening', () => console.log('I\'m listen'));
bdstt.on('stop', () => console.log('Copy that'));
bdstt.on('upload', () => console.log('Uploading voice data...'));
bdstt.on('timeout', () => console.log('TimeOUT'))
bdstt.on('sleep', () => console.log('I\'m sleeping...'))
bdstt.on('wakeup', (word) => console.log('You wake me up with ', word));

bdstt.on('success', data => {
  console.log(data);
})

bdstt.on('fail',err=> {
  console.log(err)
})


token.js
/**
 * Baidu token
 * @author aokihu aokihu@gmail.com
 * @github https://github.com/aokihu/Baidu_STT
 * @license MIT
 * @version 0.0.1
 */

const fs = require('fs');
const EventEmitter = require('events');
const path = require('path');
const querystring = require('querystring');
// const got = require('./request.js');
const got = require('little-fetch')

// Token url
const AccessUrl = 'http://openapi.baidu.com/oauth/2.0/token';
const GRANT_TYPE = 'client_credentials';
// Session file
const SessionFile = path.resolve(__dirname, './session.json');
// Session Outdate
const TokenDeadline = 12 * 3600 * 1000;

class Token extends EventEmitter {
  constructor({ apiKey, secretKey }) {
    super();

    this.apiKey = apiKey;
    this.secretKey = secretKey;
  }

  initToken() {
    return new Promise((resolve, reject) => {
      this.vaildToken()
        .then(this.tokenReady.bind(this))
        .catch(() => {
          const params = {
            grant_type: GRANT_TYPE,
            client_id: this.apiKey,
            client_secret: this.secretKey,
          };

          this.requestToken(params)
            .then(this.tokenReady.bind(this))
            .catch(reject);
        });
    });
  }

  tokenReady(token) {
    this.emit('ready', token);
  }
  vaildToken() {
    return new Promise((resolve, reject) => {
      fs.access(SessionFile, fs.R_OK, (err) => {
        if (err) {
          // TODO: 如果没有Session文件，那么返回false
          reject();
        } else {
          // TODO: 如果有Session文件，然后检查是否已经过期
          const { token, timestamp } = JSON.parse(fs.readFileSync(SessionFile));
          const delta = new Date() - new Date(timestamp);
          delta > TokenDeadline ? reject() : resolve(token);
        }
      });
    });
  }

  requestToken(params) {
    // 从百度获取token session
    const url = `${AccessUrl}?${querystring.stringify(params)}`;
    console.log(url)
    return got({url})
      .then((body) => {
        const { access_token: token } = JSON.parse(body);
        return this.saveToken(token);
      })
      .catch(console.error);
  }

  saveToken(token) {
    return new Promise((resolve, reject) => {
      fs.writeFile(
        SessionFile,
        JSON.stringify({ token, timestamp: new Date() }),
        (err) => {
          if (err) {
            reject(err);
          } else {
            resolve(token);
          }
        }
      );
    });
  }


  /**
   * @function clearToken()
   * @description Delete token session file
   */
  clearToken() {

  }
}


module.exports = Token;


index.js
/**
 * Baidu Speech To Text
 * @author aokihu aokihu@gmail.com
 * @github https://github.com/aokihu/Baidu_STT
 * @license MIT
 * @version 2.0.3
 */

const EventEmitter = require('events');
const http = require('http');
const path = require('path');
const fs = require('fs');
const Mic = require('mic');
const {Detector,Models} = require('snowboy');
const fetch = require('little-fetch')
const Token = require('./token.js');

const MIN_REACORD_FRAME_BLOCK = 3;
const MAX_BUFFER_BLOCK_NUMBER = 32;
const MAX_BUFFER_SIZE = 8192 * MAX_BUFFER_BLOCK_NUMBER;
const BDServiceAPIUrl = 'http://vop.baidu.com/server_api/';

/**
 * Define statusl constant
 */
const STAT_NULL = 0;
const STAT_READY = 1;
const STAT_SLEEP = 2;
const STAT_WAKE_UP = 3;
const STAT_LISTENING = 4;
const STAT_UPLOAD = 5;
const STAT_PEND_UPLOAD = 6;

/**
 * @class BaiduSTT
 */
class BaiduSTT extends EventEmitter {
  /**
   * @constructor
   * @param {string} appId Baidu App ID, you can find it in baidu console
   * @param {string} apiKey App api key
   * @param {string} secretKey App secret key
   * @param {string} language Choose your language, default is 'zh'
   * @param {boolean} recordVoice Whether save the voice
   * @param {string} hotwords Hot word
   * @param {string} sensitivity Snowboy sensitivity
   * @param {string} modelFile Snowboy hot word model file
   * @param {number} gain Snowboy gain
   * @param {boolean} continual If is 'true' it will pause, 'false' is stop recording
   * @param {string} voiceRate Recoding voice rate, you can set '16000' OR '8000', but '16000' maybe is the only rate with my test
   * @param {string} voicePath Voice saved path, default is current folder
   * @param {string} voiceType Voice type, default is 'wav'
   * @param {boolean} debug Show debug information
   * @param {number} silenceDelay The delay of silence after speak command, default is 2s
   */
  constructor({
    appId,
    apiKey,
    secretKey,
    language='zh',
    recordVoice = false,
    debug = false,
    continual = false,
    modelFile = './resources/Alexa.pmdl',
    sensitivity = '0.5',
    gain = 2.0,
    silenceDelay = 2000,
    hotwords = "Alexa",
    voicePath = './',
    voiceRate = '16000',
    voiceType = 'wav' }) {

    // Call parent constrcutor function
    super();

    //
    // this._ is private property
    //
    this._ = {
      token: '', /* Baidu service session token */
      status: STAT_NULL, /* the status of class, [none, ready, requesting] */
      voiceRate, /* the voice record rate */
      voicePath, /* the voice saved path */
      voiceType, /* the voice saved file type, default is 'pcm' */
      apiKey, /* Baidu service app key, you should fetch it from your baidu console */
      secretKey, /* Baidu service app secret key, you should fetch it from your baidu console */
      appId, /* It's options */
      continual, /* PAUSE OR STOP after recording */
      modelFile,
      hotwords,
      sensitivity,
      debug,
      silenceDelay,
      exitOnSilence: 3, /* Mic delay emit silence signal */
      canStop: false, /* It can stop record */
      buffer: {
        size: 0,
        point: [],
      },
      cuid: 'aokihu_client_' + Date.now(),
      lan: language,
      record: recordVoice
    };

    // Use continal model,
    // you can speak hot word to wake up
    if(this._.continual) {
      // Snowboy Model
      this.model = new Models();
      this.model.add({
        file: this._.modelFile,
        sensitivity: this._.sensitivity,
        hotwords : this._.hotwords
      });

      // Snowboy Detector
      this.detector = new Detector({
        resource: "./resources/common.res",
        models: this.model,
        audioGain: this._.gain
      });

      this._.exitOnSilence = 0;
    }

    // Init mic
    this.mic = Mic({
      rate: this._.voiceRate,
      channels: 1,
      debug: this._.debug,
      exitOnSilence: this._.exitOnSilence,
      fileType: this._.voiceType,
    });
  }

  /**
   * @public
   * @function init()
   * @description after call this function, it will check local token session,
   *              if the token is null or outdate, it will request new token
   *              from baidu service
   */
  init() {
    const tokenSession = new Token({ apiKey: this._.apiKey, secretKey: this._.secretKey });
    tokenSession.once('ready', (token) => {
      this._.token = token;
      this._.status = 'ready';
      this.emit('ready');
      this._.status = STAT_READY; // SET STATUS = READY
    });
    tokenSession.initToken();
  }

  /**
   * @public
   * @function listen
   * @description Listen for your voice
   */
  listen() {

    const micStream = this.mic.getAudioStream();

    // Save to file
    if(this._.record){
      this._recordToFile(micStream);
    }

    // Set snowboy to listen hot wod
    if(this._.continual){
      this.detector.on('hotword', this._wakeup.bind(this));
      this.detector.on('sound', this._listening.bind(this));
      this.detector.on('silence', this._afterRecord.bind(this));
      micStream.pipe(this.detector);
      this._.status = STAT_SLEEP;
      this.emit('sleep');
    } else {
      micStream.on('startComplete', this._afterStart.bind(this));
      micStream.on('data', this._recording.bind(this));
      micStream.on('silence', this._afterRecord.bind(this));
      this._.status = STAT_LISTENING;
    }

    // start recording
    this.mic.start();
  }

  /**
   * @public
   * @function pause()
   * @description Pause mic record
   */
  pause(){
    this.mic.pause();
  }

  /**
   * @public
   * @function resume()
   * @description Resume mic record
   */
  resume(){
    this.mic.resume();
  }

  /**
   * @public
   * @function stop()
   * @description Stop mic record
   */
  stop() {
    this.mic.stop();
    this._.buffer.point = [];
    this._.buffer.size = 0;
  }

  /**
   * @private
   * @function _recordToFile()
   * @param {Stream} inputStream mic stream
   * @description Save voice wave data to file
   */
  _recordToFile(inputStream){
    const filename = `output.${this._.voiceType}`;
    const outputFileStream = fs.createWriteStream(path.resolve(this._.voicePath, filename));
    inputStream.pipe(outputFileStream);
  }

  /**
   * @private
   * @function _wakeup
   * @param {number} index
   * @param {string} hotword
   * @param {Buffer} buffer
   */
  _wakeup(index, hotword, buffer){
    this.emit('wakeup', hotword);
    this._.status = STAT_LISTENING;
    setTimeout(() => { this._.canStop = true; }, this._.silenceDelay);
  }

  /**
   * @private
   * @function _afterStart()
   * @description it will delay 3s to stop record
   */
  _afterStart() {
    this.emit('start');
    this.emit('begin');
    setTimeout(() => { this._.canStop = true; }, this.silenceDelay);
  }

  /**
   * @private
   * @function _recording(data)
   * @param {Buffer} data voice data
   * @description Allocate memory and save voice data
   */
  _recording(data) {

    // Check buffer size
    // if size is lager than MAX_BUFFER_SIZE
    // free buffer and set buffer size to zero
    if(MAX_BUFFER_SIZE > this._.buffer.size)
    {
      this._.buffer.size += data.byteLength;
      this._.buffer.point.push(data);
    }else{
      this._.buffer.point = [];
      this._.buffer.size = 0;
      this.emit('timeout');
      this._.status = STAT_SLEEP;
    }

    this.emit('listening');
  }

  /**
   * @private
   * @param {Buffer} buffer voice buffer data
   */
  _listening(buffer){
    if(this._.status === STAT_LISTENING){
      this._recording(buffer);
    }
  }

  /**
   * @private
   * @function do something after record
   */
  _afterRecord() {

    if (this._.canStop) {
      if(this._.continual){
        if(this._.status === STAT_LISTENING){
          this._.status = STAT_PEND_UPLOAD;
        } else {
          return 0;
        }
      }
      else{
        this.mic.stop();
        this._.status = STAT_SLEEP;
        this.emit('stop');
      }

      const buffer = Buffer.concat(this._.buffer.point, this._.buffer.size);

      if(this._.buffer.size > 8192 * MIN_REACORD_FRAME_BLOCK)
      {
        this._uploadVoice(buffer);
      }

      this._.canStop = false;

      // Free buffer
      this._.buffer.point = [];
      this._.buffer.size = 0;
    }

  }

  /**
   * @private
   * @function uploadVoice()
   * @param {buffer} data voice data
   * @description Upload voice date to baidu STT service
   */
  _uploadVoice(data) {

    const url = `http://vop.baidu.com/server_api?token=${this._.token}&cuid=${this._.cuid}&lan=${this._.lan}`;

    fetch({
      url,
      method: 'POST',
      headers: {
        'Content-Type':`audio/wav;rate=${this._.voiceRate}`
      },
      postData:data
    })
    .then(data => {
      const parsedData = JSON.parse(data);
      if(parsedData.err_no === 0){
        this.emit('success', parsedData.result);
      }else
      {
        this.emit('fail', parsedData.err_msg);
      }

      // Set status to SLEEP
      if(this._.continual)
      {
        this._.status = STAT_SLEEP;
        this.emit('sleep');
      }

    })
    .catch(console.log)

    this.emit('upload');

  }

}


module.exports = BaiduSTT;
